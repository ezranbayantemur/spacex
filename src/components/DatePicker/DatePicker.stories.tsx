import React from 'react';
import {Text, View} from 'react-native';
import {format} from 'date-fns';
import {color, font, padding, typography} from '@style';
import DatePicker from './DatePicker';
import type {Meta, StoryObj} from '@storybook/react-native';
import type {DatePickerProps} from './DatePicker.types';

const DatePickerMeta: Meta<DatePickerProps> = {
  argTypes: {
    onDateSelected: {
      action: 'on selection',
    },
  },
  title: 'DatePicker',
  component: DatePicker,
  decorators: [
    Story => (
      <View style={{flex: 1, backgroundColor: color.BLACK}}>
        <Story />
      </View>
    ),
  ],
};

export default DatePickerMeta;

export const Basic: StoryObj<DatePickerProps> = {};

export const DateSelection: StoryObj<DatePickerProps> = {
  decorators: [
    () => {
      const [date, setDate] = React.useState<Date>(new Date());

      return (
        <View>
          <Text
            style={{
              color: color.ALABASTER,
              fontFamily: font.BOLD,
              fontSize: typography.MEDIUM,
              margin: padding.MEDIUM,
            }}>
            Selected Date:{' '}
            <Text
              style={{
                fontFamily: font.MEDIUM,
              }}>
              {format(date, 'dd MMMM yyyy')}
            </Text>
          </Text>
          <DatePicker onDateSelected={setDate} />
        </View>
      );
    },
  ],
};
