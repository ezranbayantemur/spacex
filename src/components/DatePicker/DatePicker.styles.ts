import {StyleSheet} from 'react-native';
import {color, font, padding, typography} from '@style';

const base = StyleSheet.create({
  container: {
    justifyContent: 'center',
    alignItems: 'center',
    padding: padding.SMALL,
    margin: padding.SMALL,
    borderWidth: 2,
  },
  date: {
    fontFamily: font.BOLD,
    fontSize: typography.MEDIUM,
  },
});

export default {
  default: {
    container: {
      ...base.container,
      borderColor: color.ALABASTER,
    },
    date: {
      ...base.date,
      color: color.ALABASTER,
    },
  },
  dirty: {
    container: {
      ...base.container,
      backgroundColor: color.SMOKE,
      borderColor: 'transparent',
    },
    date: {
      ...base.date,
      color: color.BLACK,
    },
  },
};
