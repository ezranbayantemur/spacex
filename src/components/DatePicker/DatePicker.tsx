import React from 'react';
import {Text, TouchableOpacity} from 'react-native';
import RNDatePicker from 'react-native-date-picker';
import {format} from 'date-fns';
import styles from './DatePicker.styles';

import type {DatePickerProps, DatePickerStyleType} from './DatePicker.types';

const DatePicker = ({onDateSelected, testID, ...rest}: DatePickerProps) => {
  const [date, setDate] = React.useState(new Date());
  const [isOpen, setIsOpen] = React.useState(false);
  const [isDirty, setIsDirty] = React.useState(false);

  const styleType: DatePickerStyleType = isDirty ? 'dirty' : 'default';

  const handleConfirm = (date: Date) => {
    setIsDirty(true);
    setIsOpen(false);
    setDate(date);
    onDateSelected(date);
  };

  const handleCancel = () => {
    setIsOpen(false);
  };

  return (
    <TouchableOpacity
      testID={`${testID}_touchable`}
      style={styles[styleType].container}
      onPress={() => setIsOpen(true)}>
      <Text style={styles[styleType].date}>{format(date, 'dd MMM yyyy')}</Text>
      <RNDatePicker
        {...rest}
        testID={`${testID}_picker`}
        modal
        mode="date"
        date={date}
        open={isOpen}
        onConfirm={handleConfirm}
        onCancel={handleCancel}
      />
    </TouchableOpacity>
  );
};

export default DatePicker;
