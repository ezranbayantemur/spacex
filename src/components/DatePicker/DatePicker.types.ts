import {DatePickerProps as RNDatePickerProps} from 'react-native-date-picker';

export interface DatePickerProps
  extends Omit<
    RNDatePickerProps,
    'modal' | 'mode' | 'date' | 'open' | 'onConfirm' | 'onCancel'
  > {
  onDateSelected: (date: Date) => void;
}

export type DatePickerStyleType = 'default' | 'dirty';