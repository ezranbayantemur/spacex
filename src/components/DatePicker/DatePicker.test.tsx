import React from 'react';
import {render, fireEvent} from '@testing-library/react-native';
import DatePicker from './DatePicker';

jest.useFakeTimers().setSystemTime(new Date('2010-01-01'));
const mockOnDateSelected = jest.fn();
jest.mock('react-native-date-picker', () => 'RNDatePicker');

let wrapper: ReturnType<typeof render>;

describe('DatePicker unit tests', () => {
  beforeEach(() => {
    wrapper = render(
      <DatePicker testID="datepicker" onDateSelected={mockOnDateSelected} />,
    );
  });

  it('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should update the date label', () => {
    const picker = wrapper.getByTestId('datepicker_picker');
    fireEvent(picker, 'onConfirm', new Date('2020-01-01'));

    expect(wrapper.getByText('01 Jan 2020')).not.toBeNull();
  });

  it('should change the style when gets dirty', () => {
    const picker = wrapper.getByTestId('datepicker_picker');
    fireEvent(picker, 'onConfirm', new Date('2020-01-01'));

    const bgcolor = wrapper.getByTestId('datepicker_touchable').props.style
      .backgroundColor;

    expect(bgcolor).toBe('#BDBDBD');
  });
});
