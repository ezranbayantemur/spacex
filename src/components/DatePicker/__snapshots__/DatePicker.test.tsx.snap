// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker unit tests should render correctly 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": undefined,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "borderColor": "#FBFBFB",
      "borderWidth": 2,
      "justifyContent": "center",
      "margin": 8,
      "opacity": 1,
      "padding": 8,
    }
  }
  testID="datepicker_touchable"
>
  <Text
    style={
      {
        "color": "#FBFBFB",
        "fontFamily": "Quicksand-Bold",
        "fontSize": 17,
      }
    }
  >
    01 Jan 2010
  </Text>
  <RNDatePicker
    date={2010-01-01T00:00:00.000Z}
    modal={true}
    mode="date"
    onCancel={[Function]}
    onConfirm={[Function]}
    open={false}
    testID="datepicker_picker"
  />
</View>
`;
